File: Terceiro-Semestre/Algoritmo-e-estruturas-de-dados-II-AEDSII/tps_resolvidos/tp01/Ex11/Palindromo.c
#include <stdio.h>
#include <string.h>

int verificarPalindromoRecursivo(char *palavra, int esquerda, int direita) {

    if (esquerda >= direita) {
        return 1; // É um palíndromo
    }

    if (palavra[esquerda] != palavra[direita]) {
        return 0; // Não é um palíndromo
    }

    return verificarPalindromoRecursivo(palavra, esquerda + 1, direita - 1);
}

void verificarPalindromo(char *palavra) {
    int tamanho = strlen(palavra);

    if (verificarPalindromoRecursivo(palavra, 0, tamanho - 1)) {
        printf("SIM\n");
    } else {
        printf("NAO\n");
    }
}

int main() {
    char palavra[1000]; 
    int continuar = 1;

    while (continuar) {
        fgets(palavra, sizeof(palavra), stdin);

        palavra[strcspn(palavra, "\n")] = '\0';


        if (strcmp(palavra, "FIM") == 0) {
            continuar = 0;
        } else {
            verificarPalindromo(palavra);
        }
    }

    return 0;
}


    

File: Terceiro-Semestre/Algoritmo-e-estruturas-de-dados-II-AEDSII/tps_resolvidos/tp01/Ex10/Palindromo.java
package Ex10;

import java.util.Scanner;

class Palindromo {
    public static void VerificarPalindromo(String palavra) {
        if (isPalindromo(palavra, 0, palavra.length() - 1)) {
            System.out.println("SIM");
        } else {
            System.out.println("NAO");
        }
    }

    private static boolean isPalindromo(String palavra, int esquerda, int direita) { // Tive como base a busca binária, ele verifica se é palíndromo e vai mudando as posições "esquerda + 1" e "direita - 1"
        if (esquerda >= direita) {
            return true;
        }

        if (palavra.charAt(esquerda) != palavra.charAt(direita)) {
            return false;
        }

        return isPalindromo(palavra, esquerda + 1, direita - 1);
    }

    public static void main(String[] args) {
        Scanner leitura = new Scanner(System.in);

        String palavra;

        palavra = leitura.nextLine();

        while (!(palavra.equals("FIM"))) {
            VerificarPalindromo(palavra);
            palavra = leitura.nextLine();
        }

        leitura.close();
    }
}
